maths/fibonacci_sequence_recursion_with_memoization.py

```diff
@@ -0,0 +1,36 @@
+# Fibonacci Sequence Using Recursion and Memoization
+
+# Memoization Concept: https://en.wikipedia.org/wiki/Memoization
+# Fibonacci Concept : https://pt.wikipedia.org/wiki/Sequ%C3%AAncia_de_Fibonacci
+
+cache = dict()
+
+
+def fibonacci(n):
+    """
+    Return number of index n in sequence fibonacci
+    >>> [fibonacci(i) for i in range(12)]
+    [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]
+    """
+    aux = n
+
+    if aux in cache:
+        return cache[aux]
+
+    if (n == 0) or (n == 1):
+        ans = n
+    else:
+        ans = fibonacci(n - 1) + fibonacci(n - 2)
+    cache[aux] = ans
+
+    return ans
+
+
+def main() -> None:
+    number = int(input("Enter an integer greater than or equal to 0: "))
```

Comment:
```suggestion
    try:
        number = int(input("Enter an integer greater than or equal to 0: "))
    except ValueError:
        print("Error: Input value should be an integer greater than or equal to 0")
```

Reply:
This try does not assert that the value is greater than or equal to 0, so the message of error is misleading.